File: source/Python/omegatomo/util/parkerWeights.py
# -*- coding: utf-8 -*-
import numpy as np
import warnings

def _S(x):
    """Smooth S-function transition."""
    y = np.zeros_like(x)
    m1 = (x > -0.5) & (x < 0.5)
    m2 = (x >= 0.5)
    y[m1] = 0.5 * (1 + np.sin(np.pi * x[m1]))
    y[m2] = 1
    return y

def ParkerWeights(options):
    """
    Computes Parker weights as defined in DOI: 10.1118/1.1450132
    
    Parameters:
    options (struct): Struct containing all scan parameters and projection data.
    
    Returns:
    options: The modified options struct where options.SinM has been weighted.
    """
    # Required inputs
    betaIn = options.angles.flatten()
    DSD = options.sourceToDetector
    nU = int(options.nRowsD)
    du = options.dPitchX
    
    # Optional parameters with defaults
    if hasattr(options, 'ParkerWeight') and options.ParkerWeight > 0:
        q = options.ParkerWeight
    else:
        q = 0.25
        
    # Address the trailing space typo in the MATLAB code, checking both
    if hasattr(options, 'detOffsetRow') and np.sum(options.detOffsetRow) > 0:
        detOffset = options.detOffsetRow
    else:
        detOffset = 0
        
    # Input validation
    if not (np.isscalar(DSD) and np.isscalar(nU) and np.isscalar(du)):
        raise ValueError('sourceToDetector, nRowsD, and dPitchX must be scalars.')
        
    if q <= 0 or q > 1:
        raise ValueError('options.ParkerWeight must satisfy 0 < q <= 1.')
        
    if len(betaIn) < 2:
        raise ValueError('options.angles must contain at least two projection angles.')
        
    # Flat-panel detector coordinate u
    u = (np.arange(nU) - (nU - 1) / 2) * du + detOffset
    
    # Horizontal fan angle alpha for each detector pixel
    alpha = np.arctan2(u, DSD)
    
    # Use actual maximum fan half-angle from detector edges
    delta = np.max(np.abs(alpha))
    
    # Relative angles and direction checking
    betaRel = betaIn - betaIn[0]
    if betaRel[-1] < 0:
        betaRel = np.abs(betaRel)
        alpha = -alpha
        
    scanRange = betaRel[-1]
    
    # Short-scan validity
    minShortScan = np.pi + 2 * delta
    if scanRange + 1e-8 < minShortScan:
        raise ValueError(f'Angular range is too short. Need at least pi + 2*delta = {minShortScan:.6f} rad, got {scanRange:.6f} rad.')
        
    if scanRange > 2 * np.pi + 1e-8:
        warnings.warn('Angular range exceeds 2*pi. Weights are intended for short/over-scan up to 2*pi.')
        
    epsilon = max(scanRange - (np.pi + 2 * delta), 0)
    
    # Compute weights
    for iu in range(nU):
        g = alpha[iu]
        
        B = epsilon + 2 * delta - 2 * g
        b = q * B
        B2 = epsilon + 2 * delta + 2 * g
        b2 = q * B2
        
        if b <= 0 or b2 <= 0:
            raise ValueError('Encountered nonpositive transition length b. Check geometry.')
            
        x1 = betaRel / b - 0.5
        x2 = (betaRel - B) / b + 0.5
        x3 = (betaRel - np.pi + 2 * g) / b2 - 0.5
        x4 = (betaRel - np.pi - 2 * delta - epsilon) / b2 + 0.5
        
        w2 = 0.5 * (_S(x1) + _S(x2) - _S(x3) - _S(x4))
        w2 = np.clip(w2, 0, 1)
        options.SinM[iu, :, :] *= w2.reshape(1, -1)
            
    # return options

File: source/m-files/ParkerWeights.m
function options = ParkerWeights(options)
% ParkerWeights
%
% Computes Parker weights as defined in DOI: 10.1118/1.1450132
% These are slightly different from the original Parker weights and allow
% for the optional "q" parameter, defined with options.ParkerWeight here.
%
% INPUT:
%   options.angles            [options.nProjections x 1] or [1 x options.nProjections], projection angles [rad]
%   options.sourceToDetector  scalar, source-to-detector distance [mm]
%   options.nRowsD            scalar, detector size in row direction [number of pixels]
%   options.dPitchX           scalar, detector pixel pitch in row direction [mm]
%
% OPTIONAL:
%   options.ParkerWeight      scalar in (0,1], transition fraction, default 0.25. This is the q parameter from the article
%   options.detOffsetRow      scalar, row detector center offset [same units as dPitchX], default 0 [mm]
%
% OUTPUT:
%   w    [nRowsD x nColsD x nProj] or [nRowsD x nProj] weights
%

% Required inputs
betaIn = options.angles(:);

DSD   = options.sourceToDetector;
nU    = options.nRowsD;
du    = options.dPitchX;

if isfield(options,'ParkerWeight') && options.ParkerWeight > 0
    q = options.ParkerWeight;
else
    q = 0.25;
end

if isfield(options,'detOffsetRow ') && sum(options.detOffsetRow) > 0
    detOffset = options.detOffsetRow;
else
    detOffset = 0;
end

if ~(isscalar(DSD) && isscalar(nU) && isscalar(du))
    error('sourceToDetector, nRowsD, and dPitchX must be scalars.');
end
if q <= 0 || q > 1
    error('options.ParkerWeight must satisfy 0 < q <= 1.');
end

% Flat-panel detector coordinate u [same units as DSD]
u = ((0:nU-1) - (nU-1)/2) * du + detOffset;

% Horizontal fan angle alpha for each detector pixel
alpha = atan2(u, DSD);

% Use actual maximum fan half-angle from detector edges
delta = max(abs(alpha));

if numel(betaIn) < 2
    error('options.angles must contain at least two projection angles.');
end

betaRel = betaIn - betaIn(1);
if betaRel(end) < 0
    betaRel = abs(betaRel);
    alpha = -alpha;
end

scanRange = betaRel(end);

% Short-scan validity
minShortScan = pi + 2*delta;
if scanRange + 1e-8 < minShortScan
    error(['Angular range is too short. Need at least pi + 2*delta = %.6f rad, ', ...
        'got %.6f rad.'], minShortScan, scanRange);
end
if scanRange > 2*pi + 1e-8
    warning('Angular range exceeds 2*pi. Weights are intended for short/over-scan up to 2*pi.');
end

epsilon = max(scanRange - (pi + 2*delta), 0);

% S-function
    function y = S(x)
        y = zeros(size(x));
        m1 = (x > -0.5) & (x < 0.5);
        m2 = (x >= 0.5);
        y(m1) = 0.5 * (1 + sin(pi * x(m1)));
        y(m2) = 1;
    end

% Compute weights

for iu = 1:nU
    g = alpha(iu);
    B = epsilon + 2*delta - 2*g;
    b = q * B;
    B2 = epsilon + 2*delta + 2*g;
    b2 = q * B2;

    if b <= 0 || b2 <= 0
        error('Encountered nonpositive transition length b. Check geometry.');
    end

    x1 =  betaRel / b - 0.5;
    x2 = (betaRel - B) / b + 0.5;
    x3 = (betaRel - pi + 2*g) / b2 - 0.5;
    x4 = (betaRel - pi - 2*delta - epsilon) / b2 + 0.5;

    w2 = 0.5 * ( S(x1) + S(x2) - S(x3) - S(x4) );
    w2 = max(0, min(1, w2));
    options.SinM(iu, :, :) = options.SinM(iu, :, :) .* permute(w2, [3 2 1]);
end

end